assayist/client/modify.py

```diff
@@ -0,0 +1,106 @@
+    WHERE {}
+    RETURN c
+    """.format(c_type, c_namespace, ' OR '.join(names_where_clause))
+
+    results, _ = neomodel.db.cypher_query(query)
+    components = [Component.inflate(row[0]) for row in results]
+
+    # If no matching component is returned, just create one
+    if not components:
+        Component(canonical_namespace=c_namespace, canonical_name=c_name, canonical_type=c_type,
+                  alternative_names=alternatives).save()
+        return
+
+    for c in components:
+        # If this property is null in Neo4j, Neomodel doesn't treat it as an empty list, so just set
+        # it here to make the code below simpler since it can always assume it's a list
+        if c.alternative_names is None:
+            c.alternative_names = []
+
+    # This will be the only remaining component if there is more than one component returned, as
+    # the information stored in the others will be merged into this one
+    component = components[0]
+    if component.canonical_name != c_name:
+        # Add the old canonical name as an alternative name
+        if component.canonical_name not in component.alternative_names:
+            component.alternative_names.append(component.canonical_name)
+        # Remove the new canonical name from the alternative names if it's present
+        if c_name in component.alternative_names:
+            component.alternative_names.pop(component.alternative_names.index(c_name))
+        # Set the proper canonical name
+        component.canonical_name = c_name
+
+    # Add the missing alternative names
+    for alternative in alternatives:
+        if alternative not in component.alternative_names:
+            component.alternative_names.append(alternative)
+
+    # If there was only one component returned, then it's all set and no merging is required
+    if len(components) == 1:
+        component.save()
+        return
+
+    # If there was more than one component that matched, we must merge them
+    for c in components[1:]:
+        # Merge all the source locations on the first component
+        for sl in c.source_locations.all():
+            component.source_locations.connect(sl)
+
+        # Add the canonical name of this component as an alternative name for the first component,
+        # if applicable
+        if c.canonical_name != component.canonical_name and \
+                c.canonical_name not in component.alternative_names:
+            component.alternative_names.append(c.canonical_name)
+        # Add all the alternative names of this component to the first component, if applicable
+        for alternative in c.alternative_names:
+            if alternative != component.canonical_name and \
+                    alternative not in component.alternative_names:
+                component.alternative_names.append(alternative)
+        # Delete this component since all its information is now stored in the first component
+        c.delete()
```

Comment:
As mentioned in today's standup, should we log the component that is being removed so we have an audit trail in case we need to revert some changes?

Reply:
Where would we log that?

Reply:
`log.info()` I guess. That should still be shown in the Jenkins jobs, no?

Reply:
The user would be calling this particular code, and not Jenkins. So I'll just use a print statement instead.

Reply:
Ah, that's right, I forgot this is part of the client lib. `log.warn` makes sense. Thanks!